### PR TITLE
adding clean command

### LINKED ISSUE #102: Clean command
We should add a command that will recursively remove the `dest` directory.

### PR DESCRIPTION
**Pull request by rjgoldsborough**
fixes #102 

If there is a cleaner way to do this, please let me know.
I kept getting `expected (), found enum 'std::result::Result\'` but writing it this way worked.

I did not add a test for this just yet but trying to figure that out.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -258,6 +260,13 @@ fn main() {
             }
         }
 
+        "clean" => {
+            match fs::remove_dir_all(&config.dest) {
```
Comment (nott) [inline src/main.rs]: `cobalt build` doesn't create `dest` directory:
```
$ cobalt build
[info]   Using config file ./.cobalt.yml
[info]   Building from . into build
[error]  IO error: No such file or directory (os error 2)
[error]  Build not successful
```

I think it meanse `cobalt clean` shouldn't remove it.

File: src/main.rs [hunk 2]
```diff
@@ -258,6 +260,13 @@ fn main() {
...
+              Ok(..) => println!("directory \"{}\" removed", &config.dest),
+              Err(err) => println!("Error: {}", err)
```
Comment (nott) [inline src/main.rs]: could you please use `error!` instead of `println!` here?

**Comment by nott** [comment]
Travis build is failing due to issues with formatting.

You can fix them with [cargo fmt](https://github.com/rust-lang-nursery/rustfmt#quick-start)

**Comment by rjgoldsborough** [comment]
Thanks for the feedback. (sorry for the sloppiness :))
I will incorporate these.

File: src/main.rs (hunk 1 above)
Comment (nott) [inline src/main.rs]: Makes sense indeed.

Comment (rjgoldsborough) [inline src/main.rs]: I'll definitely give it a shot and come back with any questions or issues I run into.

**Comment by johannhof** [comment]
Thanks for the PR! What you *could* do to test this is to take a look at [tests/cli.rs](https://github.com/cobalt-org/cobalt.rs/blob/master/tests/cli.rs) and add a new test for `clean` that runs

```
env::set_current_dir(CWD.join("tests/fixtures/example")).unwrap();
assert_cli!(&BIN, &["build", "-d", "./test_dest"] => Success).unwrap();
```
check if  `./test_dest` was created and then run 

```
let output = assert_cli!(&BIN, &["clean", "-d", "./test_dest"] => Success).unwrap();
```

and check if `./test_dest` was removed. You should also check if `output` has the correct messages.

Let me know how that works for you :)

Thanks!

File: src/main.rs (hunk 1 above)
Comment (rjgoldsborough) [inline src/main.rs]: hmm weird. cobalt *does* create the dir for me.

https://gist.github.com/rjgoldsborough/08fc7700a847b9318a7ebd1da10ade89

**Comment by rjgoldsborough** [comment]
@johannhof, thanks for the suggestions.

It seems to have worked for the most part. I was trying to check the output but was getting errors about `PartialEq` and `&str` so I pushed the rest of the test until I get that sorted.

**Comment by rjgoldsborough** [comment]
One other question I had with that last commit was if we should use `create_dir_all` for the case of a user supplying a multi level dest dir?

File: src/main.rs (hunk 1 above)
Comment (nott) [inline src/main.rs]: @rjgoldsborough 

my comment was misleading, sorry for that
i tested on version 0.3, latest master creates dest dir automatically indeed

File: src/cobalt.rs [hunk 3]
```diff
@@ -42,6 +42,13 @@ pub fn build(config: &Config) -> Result<()> {
     let dest = PathBuf::from(&config.dest).join("");
     let dest = dest.as_path();
 
+    // try to create the supplied dest dir if it currently does not exist
+    if dest.is_dir() == false {
+        try!(fs::create_dir(dest));
+        info!("Created dest dir: {:?}", dest);
+    }
```
Comment (johannhof) [inline src/cobalt.rs]: I'm pretty sure you can remove this again. With your new test it should be easy to verify. Sorry for the confusion. :)

File: tests/cli.rs [hunk 4]
```diff
@@ -65,3 +65,12 @@ pub fn log_levels() {
...
 }
 
+#[test]
+pub fn clean() {
+    env::set_current_dir(CWD.join("tests/fixtures/example")).unwrap();
+    assert_cli!(&BIN, &["build", "-d", "./test_dest"] => Success).unwrap();
+    assert_eq!(Path::new("./test_dest/").is_dir(), true);
+
+    assert_cli!(&BIN, &["clean", "-d", "./test_dest"] => Success).unwrap();
+    assert_eq!(Path::new("./test_dest").is_dir(), false);
```
Comment (johannhof) [inline tests/cli.rs]: why can't you use `assert_contains` like in the above examples here? E.g.:

```rust
assert_contains!(&output.stderr, "directory \"./test_dest/\" removed");
```

**Comment by rjgoldsborough** [comment]
I removed that code to add the dest dir but had an issue trying to test the output.

It seems like there is no default config file when testing so the output when running clean was:

```
running 3 tests
test invalid_calls ... ok
test log_levels ... ok
test clean ... FAILED

failures:

---- clean stdout ----
	[warn]   No .cobalt.yml file found in current directory, using default config.
[info]   directory "./test_dest" removed

thread 'clean' panicked at 'assertion failed: text.contains("[info]    directory \"./test_dest\" removed")', tests/cli.rs:77
```

so I think maybe that warn is getting tested as well?

Any ideas on how to handle that?

**Comment by johannhof** [comment]
Hm maybe it's because of the whitespace between `[info]` and the rest?

For me adding

```rust
assert_contains!(&output.stderr, "directory \"./test_dest\" removed");
```

passes all tests :)

**Comment by rjgoldsborough** [comment]
Huh yeah, I dunno. I reran the tests this AM and they seem fine.

**Comment by johannhof** [comment]
Great, thanks!

**Comment by nott** [comment]
Awesome

**Pull request merged**